internal/storage/streamed_chunk_uploader.go

```diff
@@ -0,0 +1,206 @@
+	}
+
+	if req.GenerationPrecondition != nil && *req.GenerationPrecondition != 0 {
+		return nil, fmt.Errorf("request received for pre-existing object %s, supported only for new objects", req.Name)
+	}
+
+	if writeChunkSize <= 0 {
+		return nil, fmt.Errorf("writeChunkSize <= 0 not supported")
+	}
+
+	// raw initialization
+	sow := StreamedChunkUploader{}
+	sow.state = Uninitialized
+
+	// store references to necessary parameters
+	sow.objectName = obj.ObjectName()
+	sow.userProgressFunc = progressFunc
+
+	// Create a NewWriter with the requested attributes, using Go Storage Client.
+	// Chuck size for resumable upload is default i.e. 16MB.
+	wc := obj.NewWriter(ctx)
+	// NewWriter never returns nil, so no nil-check here.
+
+	wc = storageutil.SetAttrsInWriter(wc, req)
+	wc.ChunkSize = writeChunkSize
+	wc.ProgressFunc = func(n int64) {
+		sow.totalWriteSucceededSoFar = n
+		logger.Debugf("%d bytes copied so far for object/file %s. chunk-size = %d", n, req.Name, wc.ChunkSize)
+
+		if sow.userProgressFunc != nil {
+			sow.userProgressFunc(n)
+		}
+	}
+	sow.wc = wc
+	sow.state = Initialized
+
+	return &sow, nil
+}
+
+func (sow StreamedChunkUploader) ChunkSize() (int, error) {
+	if nil == sow.wc {
+		return 0, fmt.Errorf("writer not initialized")
+	}
+	return sow.wc.ChunkSize, nil
+}
+
+func (sow StreamedChunkUploader) State() UploaderState {
+	return sow.state
+}
+
+func (sow StreamedChunkUploader) BytesUploadInitiatedSoFar() int64 {
+	return sow.totalWriteInitiatedSoFar
+}
+
+func (sow StreamedChunkUploader) BytesUploadedSoFar() int64 {
+	return sow.totalWriteSucceededSoFar
+}
+
+func (sow *StreamedChunkUploader) readyToWrite() bool {
+	switch sow.state {
+	case Initialized:
+		fallthrough
+	case Waiting:
+		fallthrough
+	case Writing:
+		return true
+	default:
+		return false
+	}
+}
+
+func (sow *StreamedChunkUploader) readyToClose() bool {
+	switch sow.state {
+	case Initialized:
+		fallthrough
+	case Waiting:
+		fallthrough
+	case Writing:
+		fallthrough
+	case WriteError:
+		return true
+	default:
+		return false
+	}
+}
+
+// Write uploads the given content
+// to underlying GCS object with the initial configures chunk-size,
+// number of bytes, an object writer, and a content-reader.
+// It is assumed that all the relevant attributes of the object
+// writer have been already by a previous calls to CreateObject
+// which created the object-writer.
+func (sow *StreamedChunkUploader) UploadChunkAsync(contents io.Reader) error {
+	if !sow.readyToWrite() || (sow.wc == nil) {
+		return fmt.Errorf("writer not ready to write: object: %s, status = %v, writer: %v", sow.objectName, sow.state, sow.wc)
+	}
+
+	n, err := io.Copy(sow.wc, contents)
+	sow.totalWriteInitiatedSoFar += n
+	if n == 0 || err == io.EOF {
+		// for io.EOF, don't change/append the error (message) as it makes it difficult to identify EOF for the caller.
+		sow.state = Waiting
+		return io.EOF
```

Comment:
what happens if you try to write data < chunksize, will the writer wait till the next set of data comes?

Reply:
No, it doesn't wait, it just returns with io.EOF, which is a harmless error.

Reply:
When does the upload happen of that data happens? when we write the next chunk or call close?

Reply:
It happens during the next upload call(s) whenever a full chunk is available.